lib/remote/httpmessage.hpp

```diff
@@ -0,0 +1,307 @@
+		{
+			ec = {};
+			body.more = true;
+#if BOOST_VERSION >= 107000
+			if (n) {
+				body.buffer.reserve(*n);
+			}
+#endif /* BOOST_VERSION */
+		}
+
+		template <class ConstBufferSequence>
+		std::size_t put(const ConstBufferSequence& buffers, boost::beast::error_code& ec)
+		{
+			auto size = boost::asio::buffer_size(buffers);
+			if (size > body.buffer.max_size() - body.Size()) {
+				ec = boost::beast::http::error::buffer_overflow;
+				return 0;
+			}
+
+			auto const wBuf = body.buffer.prepare(size);
+			boost::asio::buffer_copy(wBuf, buffers);
+			body.buffer.commit(size);
+			return size;
+		}
+
+		void finish(boost::beast::error_code& ec)
+		{
+			ec = {};
+			body.more = false;
+		}
+
+	private:
+		value_type& body;
+	};
+
+	/**
+	 * Implement the boost::beast BodyWriter interface for this body type
+	 *
+	 * This is used (for example) by the @c boost::beast::http::serializer to write out the
+	 * message over the TLS stream. The logic is similar to the writer of the
+	 * @c boost::beast::http::buffer_body.
+	 *
+	 * On the first call, the get method will return the entire buffer and set a toggle.
+	 *
+	 * On the second call, it will free the buffer range returned previously. Also, if there
+	 * is more data expected, for example because a corresponding reader has not yet finished
+	 * filling out the body, a `need_buffer` error is returned, informing the serializer to
+	 * abort writing without finishing, which in turn leads to the outer call to
+	 * `http::async_write` calling their completion handlers with the `need_buffer` error,
+	 * notifying that more data is required for another call to `http::async_write`.
+	 */
+	class writer
+	{
+	public:
+		using const_buffers_type = typename decltype(value_type::buffer)::const_buffers_type;
+
+		template <bool isRequest, class Fields>
+		explicit writer(const boost::beast::http::header<isRequest, Fields>& h, value_type& b)
+			: body(b)
+		{
+		}
+
+		/**
+		 * This constructor is needed specifically for boost-1.66, which was the first version
+		 * the beast library was introduced and is still used on older (supported) distros.
+		 */
+		template <bool isRequest, class Fields>
+		explicit writer(const boost::beast::http::message<isRequest, SerializableBody, Fields>& msg)
+			: body(const_cast<value_type &>(msg.body()))
+		{
+		}
+
+		void init(boost::beast::error_code& ec) { ec = {}; }
+
+		boost::optional<std::pair<const_buffers_type, bool>> get(boost::beast::error_code& ec)
+		{
+			using namespace boost::beast::http;
+
+			if (sizeWritten > 0) {
+				body.buffer.consume(std::exchange(sizeWritten, 0));
+			}
+
+			if (body.buffer.size()) {
+				ec = {};
+				sizeWritten = body.buffer.size();
+				return {{body.buffer.data(), body.more}};
+			}
+
+			if (body.more) {
+				ec = {make_error_code(error::need_buffer)};
+			} else {
+				ec = {};
+			}
+			return boost::none;
+		}
+
+	private:
+		value_type& body;
+		decltype(body.buffer.size()) sizeWritten = 0;
```

Comment:
Why not just use `std::size_t` as a type? The return type of `value_type::Size()` is set to `size_t` already, so I would just use that type here as well.